engineering/specs/internationalization.md

```diff
@@ -0,0 +1,194 @@
+    <Component/>
+  {:else if activeSlot === 'on-this-span'}
+    <span>some content</span>
+  {/if}
+</RichText>
+```
+
+`RichText` would only parse the result of the translated string, i.e. it will not perform any translation within it. It would be a general-purpose component that gets a string and replaces placeholders with its slot content.
+
+The RichText component would parse the string, identify placeholders, split it, and render it as an array or arrays, render them with slots passing down the `activeSlot` property. This allows us to use both Components or DOM elements as we see fit.
+
+The `RichText` component would look something like this.
+
+```svelte
+<script>
+  export let text: string;
+  // text would contain `Welcome, DemonLord! Click {on-this-component} or {on-this-span} to continue!`
+
+  ...
+
+  $: splitText = parse(text);
+  /** The text would be parsed, and split to return something like:
+    * [
+    *   ["Welcome, DemonLord! Click ", "on-this-component", " or "],
+    *   ["", "on-this-span", " to continue!"]
+    * ]
+    */
+</script>
+
+{#each splitText as [prefix, slotname, postfix]}
+  {prefix}<slot activeSlot={slotname}/>{postfix}
+{/each}
+```
+
+**Standards for creating keys**
+
+1. Immutable strings will be used as keys for translations.
+2. All such keys will be in `camelCase`.
+
+**Translations inside the components library**
+
+Some components inside the component library that has some strings hard-coded inside them. E.g. `FileUpload.svelte`. We will refactor such components to accept those strings as props or slots (with English strings hard-coded in the component as defaults)
+
+**Plurals/Formatting(date, time, currency, etc...)**
+
+The [typesafe-i18n](https://github.com/ivanhofer/typesafe-i18n) library supports this on the client side using [formatters](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/formatters)
+
+### Server-side rendered pages
+
+Our application has a few pages being rendered by the Django templating engine. To support localization in those templates:
+
+Use the Django translations mechanism for templates and other backend-related translations including but not limited to API error messages that are shown on the UI as it is. This will lead to two separate translations file which should be fine. Both of the files(server-generated and client generated) can be uploaded to the translations service. This will be somewhat similar to i18next's namespaces approach. The benefits that we get here are two folds:
+
+1. Using the Django translations with Django templates is very straightforward and we do not need to do any JS trickery to use the npm installed typesafe-i18n version inside the templates.
+2. We also get translations for the API error messages. Also, note that most of the framework-generated error messages are already translated by the respective package. For ex: https://www.django-rest-framework.org/topics/internationalization/.
```

Comment:
I think we should probably handle translations at a higher level in the stack. The reasons are:
- To change the language for Postgres error messages you have to have the target locale installed on the underlying machine (not a given on Linux servers), and available for postgres.
- It's not easy to change the locale for errors on the fly based on, e.g., a request parameter, so the user client-side would get errors in whatever language the server is set up for. 

For errors thrown directly from Postgres, I think it makes sense to try to glean as much info as possible based on the diagnostic codes returned (if any). It's possible to determine these in the python layer, and we could then construct error messages in the appropriate language. This would be lots of work, though.

Reply:
I think it's fine not to localize Postgres error messages for now, since the effort required is high and the impact is probably low. I recommend that the spec say this is a known limitation and we can deal with it if/when it becomes a problem later.